### ISSUE TITLE
Kafka Consumers Scheduling and Scaling

### ISSUE THREAD

**Issue by pierDipi** [knative-extensions/eventing-kafka-broker#1537]
**Problem**

Currently, each Dispatcher replica instantiates a Consumer for each Trigger with the same consumer group. With an increasing number of Triggers, the Dispatcher resources need to be increased as well and there is no way to configure how many consumers for a specific Trigger we need to run. 
This proposal will allow partitioning consumers across Dispatcher replicas and fine-grained control over the parallelism of each Trigger.

Design doc: https://docs.google.com/document/d/1UktwiDyqq07MtA7pUlahEpux5CCdAsyI6k3nkQeeqXw/edit

**[Persona:](https://github.com/knative/eventing/blob/master/docs/personas.md)** 
Which persona is this feature for?
- Event Consumer

**Time Estimate (optional):** 
How many developer-days do you think this may take to resolve?

30

**Additional context (optional)**

/priority important-soon
/area broker
/area channel
/area source
/kind feature-request

**Comment by devguyio** [comment]
I know I'm coming late to the party. So please excuse me. I am worried about the consequences of choosing the "ConsumerGroup" name. 
1. It has a direct connotation with Kafka's consumer groups and its symantics.
2. If some Kafka operator (say Strimzi) decides to create a ConsumerGroup CRD for whatever reason, it'll collide with ours and will be confusing for the users.

I'm thinking of a more generic term. Maybe "DispatcherPartition" and "DispatcherPartitionGroup" or something along those lines.

**Comment by aavarghese** [comment]
@pierDipi I was able to deploy KafkaSource with consumer replicas (under and over [pod capacity](https://github.com/aavarghese/eventing-kafka-broker/blob/685ce5045ab537285408da6639b28460f3e0c98c/control-plane/pkg/reconciler/consumergroup/controller.go#L84)) and the HA scheduler with this set of predicates and priorities in the [scheduler CM](https://github.com/aavarghese/eventing-kafka-broker/blob/685ce5045ab537285408da6639b28460f3e0c98c/control-plane/config/200-controller/100-config-kafka-scheduler.yaml).

With 25 VReplicas for example,
```
  Placements:
    Pod Name:      kafka-source-dispatcher-0
    Vreplicas:     13
    Pod Name:      kafka-source-dispatcher-1
    Vreplicas:     12
  Replicas:        25
  ```
  
  But I was seeing an issue when running with priority plugin `{"Name": "AvailabilityZonePriority",}`
  
 The error caused dispatcher pod #2 to not start up - ```  Warning  FailedMount  98s (x9 over 3m46s)  kubelet            MountVolume.SetUp failed for volume "contract-resources" : configmap "kafka-source-dispatcher-1" not found```
 Investigating this...

**Comment by aavarghese** [comment]
@pierDipi the failed mount error above occurs intermittently (I noticed it a couple more times). But my next few runs with HA scheduler AND with priority `{"Name": "AvailabilityZonePriority",}` worked successfully with the replicas getting evenly spread across the three zones in my cluster.

```
Placements:
    Pod Name:      kafka-source-dispatcher-0
    Vreplicas:     9
    Pod Name:      kafka-source-dispatcher-1
    Vreplicas:     8
    Pod Name:      kafka-source-dispatcher-2
    Vreplicas:     8
```
Could this be something to do with the webhook not injecting CMs fast enough? Needs some playing with.

**Comment by aavarghese** [comment]
Noticed evictor misbehaving (just leaving here for documentation):

pod-1 marked for eviction leaving only pod-0 as "Schedulable" but pod replicas remain as "2"

```
{"level":"info","ts":"2022-02-16T16:25:53.150Z","logger":"kafka-broker-controller","caller":"state/state.go:324","msg":"cluster state info","knative.dev/pod":"kafka-controller-656f759f4b-7bl6w","NumPods":"2","NumZones":"3","NumNodes":"3","Schedulable":"[0 1]"}
{"level":"info","ts":"2022-02-16T16:25:53.154Z","logger":"kafka-broker-controller","caller":"statefulset/autoscaler.go:134","msg":"checking adapter capacity","knative.dev/pod":"kafka-controller-656f759f4b-7bl6w","pending":0,"replicas":2,"last ordinal":1}
{"level":"info","ts":"2022-02-16T16:25:53.168Z","logger":"kafka-broker-controller","caller":"consumergroup/evictor.go:116","msg":"Marked pod as unschedulable","knative.dev/pod":"kafka-controller-656f759f4b-7bl6w","component":"evictor","kafka.eventing.knative.dev/component":"evictor","consumergroup":"default/986e73d7-3a6a-4a9f-bf94-6e7c21bae216","pod":"knative-eventing/kafka-source-dispatcher-1"}
{"level":"info","ts":"2022-02-16T16:27:33.581Z","logger":"kafka-broker-controller","caller":"state/state.go:324","msg":"cluster state info","knative.dev/pod":"kafka-controller-656f759f4b-7bl6w","NumPods":"2","NumZones":"3","NumNodes":"3","Schedulable":"[0]"}
{"level":"info","ts":"2022-02-16T16:27:33.584Z","logger":"kafka-broker-controller","caller":"statefulset/autoscaler.go:134","msg":"checking adapter capacity","knative.dev/pod":"kafka-controller-656f759f4b-7bl6w","pending":0,"replicas":2,"last ordinal":1}
```

**Comment by aavarghese** [comment]
Another observation with consumers for 2 different KafkaSource objects:

```Ansus-MBP:eventing-kafka-broker ansuvarghese$ k get kafkasource
NAME             TOPICS                   BOOTSTRAPSERVERS                            READY   REASON   AGE
kafka-source     ["knative-demo-topic"]   ["my-cluster-kafka-bootstrap.kafka:9092"]   True             29m
kafka-source-2   ["knative-demo-topic"]   ["my-cluster-kafka-bootstrap.kafka:9092"]   True             24m

Ansus-MBP:eventing-kafka-broker ansuvarghese$ k get consumergroup
NAME                                   READY   REASON   SUBSCRIBER                                       REPLICAS   READY REPLICAS   AGE
2568fde3-d74e-4baa-b40d-b561c58744d7   True             http://event-display.default.svc.cluster.local   5          5                29m
7b184904-5551-44ac-a88e-2003ab251f69   True             http://event-display.default.svc.cluster.local   10         0                24m

Ansus-MBP:eventing-kafka-broker ansuvarghese$ k get consumer
NAME                                         READY   REASON   SUBSCRIBER                                       AGE
2568fde3-d74e-4baa-b40d-b561c58744d7-76gp6   True             http://event-display.default.svc.cluster.local   30m
```

Single consumer and Ready Replicas for second consumer group is 0
Lack of selector labels the issue?

**Comment by aavarghese** [comment]
When a consumer is removed and its binding dispatcher pod also gone (autoscaled down), the config map isn't cleaned up cc: @pierDipi

**Comment by aavarghese** [comment]
New observation # 2: The other dispatcher pods for broker and channel are also autoscaled up/down with KafkaSource and stay in `ContainerCreating` state...

```
k get pods -n knative-eventing
NAME                                      READY   STATUS              RESTARTS   AGE
kafka-broker-dispatcher-0                 0/1     ContainerCreating   0          8m30s
kafka-broker-dispatcher-1                 0/1     ContainerCreating   0          8m30s
kafka-broker-dispatcher-2                 0/1     ContainerCreating   0          6m49s
kafka-broker-dispatcher-3                 0/1     ContainerCreating   0          6m49s
kafka-broker-dispatcher-4                 0/1     ContainerCreating   0          5m9s
kafka-channel-dispatcher-0                0/1     ContainerCreating   0          8m30s
kafka-channel-dispatcher-1                0/1     ContainerCreating   0          8m30s
kafka-channel-dispatcher-2                0/1     ContainerCreating   0          6m49s
kafka-channel-dispatcher-3                0/1     ContainerCreating   0          6m49s
kafka-channel-dispatcher-4                0/1     ContainerCreating   0          5m9s
kafka-source-dispatcher-0                 1/1     Running             0          9m40s
kafka-source-dispatcher-1                 1/1     Running             0          9m40s
kafka-source-dispatcher-2                 1/1     Running             0          7m30s
kafka-source-dispatcher-3                 1/1     Running             0          7m30s
kafka-source-dispatcher-4                 1/1     Running             0          7m29s
```

**Comment by aavarghese** [comment]
New observation # 3: When eviction happens to compress low occupied higher ordinal pods and consumers need to be finalized (as pods get evicted), they stay stuck in `BindInProgress` state...

```
k get kafkasource; k get consumergroup; k get consumer
NAME             TOPICS                   BOOTSTRAPSERVERS                            READY   REASON             AGE
kafka-source-3   ["knative-demo-topic"]   ["my-cluster-kafka-bootstrap.kafka:9092"]   False   FinalizeConsumer   10m
NAME                                   READY   REASON             SUBSCRIBER                                         REPLICAS   READY REPLICAS   AGE
6d2966dd-afc3-4b48-88e1-18ae4067b072   False   FinalizeConsumer   http://event-display-2.default.svc.cluster.local   10         10               10m
NAME                                         READY   REASON           SUBSCRIBER                                         AGE
6d2966dd-afc3-4b48-88e1-18ae4067b072-2bblj   False   BindInProgress   http://event-display-2.default.svc.cluster.local   10m
6d2966dd-afc3-4b48-88e1-18ae4067b072-5t2lw   False   BindInProgress   http://event-display-2.default.svc.cluster.local   10m
6d2966dd-afc3-4b48-88e1-18ae4067b072-9zr2l   False   BindInProgress   http://event-display-2.default.svc.cluster.local   10m
6d2966dd-afc3-4b48-88e1-18ae4067b072-gnjks   True                     http://event-display-2.default.svc.cluster.local   10m
```

**Comment by pierDipi** [comment]
> New observation # 1: When a consumer is removed and its binding dispatcher pod also gone (autoscaled down), the config map isn't cleaned up cc: @pierDipi
> 
> ```
> kafka-source-dispatcher-0                 1/1     Running             0          9m40s
> kafka-source-dispatcher-1                 1/1     Running             0          9m40s
> kafka-source-dispatcher-2                 1/1     Running             0          7m30s
> kafka-source-dispatcher-3                 1/1     Running             0          7m30s
> kafka-source-dispatcher-4                 1/1     Running             0          7m29s
> ```
> 
> ```
> k get cm -n knative-eventing
> kafka-source-dispatcher-0              1      15h
> kafka-source-dispatcher-1              1      15h
> kafka-source-dispatcher-2              1      8m35s
> kafka-source-dispatcher-3              1      8m35s
> kafka-source-dispatcher-4              1      7m6s
> kafka-source-dispatcher-5              1      7m6s
> kafka-source-dispatcher-6              1      7m6s
> ```

See https://github.com/knative-sandbox/eventing-kafka-broker/pull/1932

**Comment by aavarghese** [comment]
New error preventing config maps from being created and hence dispatcher pod replicas stay in "ContainerCreating"

To reproduce: In openshift `default` project, install a new KafkaSource in `default` NS. 

```{"level":"info","ts":"2022-03-18T18:47:09.385Z","logger":"kafka-broker-controller.event-broadcaster","caller":"record/event.go:282","msg":"Event(v1.ObjectReference{Kind:\"Consumer\", Namespace:\"default\", Name:\"56ac1b53-cd59-456a-a19c-9666da173e63-bcc2r\", UID:\"e55fae12-73dc-451e-b76f-66c583d23c2e\", APIVersion:\"internal.kafka.eventing.knative.dev/v1alpha1\", ResourceVersion:\"24002544\", FieldPath:\"\"}): type: 'Warning' reason: 'InternalError' failed to bind resource to pod: failed to get or create data plane ConfigMap knative-eventing/kafka-source-dispatcher-2: configmaps \"kafka-source-dispatcher-2\" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on: , <nil>","knative.dev/pod":"kafka-controller-bfcdc9d66-mfkrh"}```

Missing some additional permissions in Cluster role for controller...cc: @pierDipi

**Comment by pierDipi** [comment]
@aavarghese do you this commit? https://github.com/knative-sandbox/eventing-kafka-broker/commit/cc53b4139bd670180c133f1d19a016138503995f

**Comment by aavarghese** [comment]
I do have that commit. Still seeing the error.

**Comment by aavarghese** [comment]
For tracking here: https://github.com/knative/eventing/issues/6285

**Comment by aslom** [comment]
I am seeing occasional restarts of dispatcher pod - I do not have easy way to reproduce it (it happens after minutes of large amounts of events sent through multiple Kafka sources). What I see for dispatcher pod is `exitCode: 143` which is not very useful 

```
 containerStatuses:
  - containerID: cri-o://bde2f2466a34fad9b08c60a2811d0ce766517d31ac2ced8c524908b8cb92a118
    image: docker.io/aslom/knative-kafka-broker-dispatcher:e61457c06d40786eebe0cc95301d08b7d0a7dcc4738064c597de3ab9d20560f2
    imageID: docker.io/aslom/knative-kafka-broker-dispatcher@sha256:f001609cdf15be945c5107bf2075c7e111890a3aa48385aaf5eb22c6613374d6
    lastState:
      terminated:
        containerID: cri-o://f3bb256d7b3f5ff006e2f4e724db30c0cd9834b7dca3f963a786fd8545899a9f
        exitCode: 143
        finishedAt: "2022-04-07T20:49:07Z"
        message: |
          read-0","level":"INFO","level_value":20000,"egress.uid":"6633ea57-10f1-41e6-b8fd-df541f110b8c","resource.uid":"6633ea57-10f1-41e6-b8fd-df541f110b8c"}
          {"@timestamp":"2022-04-07T20:49:06.841Z","@version":"1","message":"Removed egress egress.uid=3b5ed361-1498-4f42-97a3-174289b7dd79 resource.uid=3b5ed361-1498-4f42-97a3-174289b7dd79","logger_name":"dev.knative.eventing.kafka.broker.dispatcher.main.ConsumerDeployerVerticle","thread_name":"vert.x-eventloop-thread-0","level":"INFO","level_value":20000,"egress.uid":"3b5ed361-1498-4f42-97a3-174289b7dd79","resource.uid":"3b5ed361-1498-4f42-97a3-174289b7dd79"}
          {"@timestamp":"2022-04-07T20:49:06.853Z","@version":"1","message":"Removed egress egress.uid=4899712e-1bfe-45d3-908c-3fda45260220 resource.uid=4899712e-1bfe-45d3-908c-3fda45260220","logger_name":"dev.knative.eventing.kafka.broker.dispatcher.main.ConsumerDeployerVerticle","thread_name":"vert.x-eventloop-thread-0","level":"INFO","level_value":20000,"egress.uid":"4899712e-1bfe-45d3-908c-3fda45260220","resource.uid":"4899712e-1bfe-45d3-908c-3fda45260220"}
          {"@timestamp":"2022-04-07T20:49:06.867Z","@version":"1","message":"Removed egress egress.uid=99472900-f4b6-4c5d-8a7a-27246751f8d2 resource.uid=99472900-f4b6-4c5d-8a7a-27246751f8d2","logger_name":"dev.knative.eventing.kafka.broker.dispatcher.main.ConsumerDeployerVerticle","thread_name":"vert.x-eventloop-thread-0","level":"INFO","level_value":20000,"egress.uid":"99472900-f4b6-4c5d-8a7a-27246751f8d2","resource.uid":"99472900-f4b6-4c5d-8a7a-27246751f8d2"}
          {"@timestamp":"2022-04-07T20:49:06.868Z","@version":"1","message":"Freed 2 thread-local buffer(s) from thread: vert.x-eventloop-thread-0","logger_name":"io.netty.buffer.PoolThreadCache","thread_name":"vert.x-eventloop-thread-0","level":"DEBUG","level_value":10000}
          {"@timestamp":"2022-04-07T20:49:06.868Z","@version":"1","message":"Freed 2 thread-local buffer(s) from thread: vert.x-eventloop-thread-0","logger_name":"io.netty.buffer.PoolThreadCache","thread_name":"vert.x-eventloop-thread-0","level":"DEBUG","level_value":10000}
        reason: Error
        startedAt: "2022-04-07T20:42:32Z"
    name: kafka-source-dispatcher
    ready: true
    restartCount: 1
    started: true
```

**Comment by pierDipi** [comment]
https://github.com/knative-sandbox/eventing-kafka-broker/pull/2068 might be related to the restart issue.

**Comment by nikt-lsq** [comment]
When this feature becomes available will it cover kafka-broker-dispatcher as well as the kafka-source-dispatcher? - in the comments above I see only references to the kafka-source-dispatcher.

**Comment by aavarghese** [comment]
@nikt-lsq (Auto)Scaling for KafkaBroker `kafka-broker-dispatcher` is currently in the works - there are PRs open for it and getting close to full support. KafkaSource scaling and scheduling is complete and has been available since release 1.6.0.
